common/predicates/predicates.go

```diff
@@ -0,0 +1,172 @@
+// The MIT License
+//
+// Copyright (c) 2020 Temporal Technologies Inc.  All rights reserved.
+//
+// Copyright (c) 2020 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package predicates
+
+import "fmt"
+
+type (
+	Predicate[T any] interface {
+		Test(T) bool
+	}
+
+	AndImpl[T any] struct {
+		Predicates []Predicate[T]
+	}
+
+	OrImpl[T any] struct {
+		Predicates []Predicate[T]
+	}
+
+	NotImpl[T any] struct {
+		Predicate Predicate[T]
+	}
+
+	AllImpl[T any] struct{}
+
+	NoneImpl[T any] struct{}
+)
+
+func And[T any](
+	predicates ...Predicate[T],
+) Predicate[T] {
+	if len(predicates) < 2 {
+		panic(fmt.Sprintf("And requires at least 2 predicates, got %v", len(predicates)))
+	}
```

Comment:
< 1?

Reply:
I think And needs at least two operands? 
If we are doing predicate1.And([]Predicates) then it should be < 1.

Reply:
i guess making <1 will make the logic simpler, up to you